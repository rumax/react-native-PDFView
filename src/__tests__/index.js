/* @flow */
import React from 'react';
import renderer from 'react-test-renderer';

import {
  findNodeHandle,
  NativeModules,
} from 'react-native';

import PDFVIew from '../index';

jest.mock('../RNPDFView', () => 'RNPDFView');
jest.mock('react-native', () => ({
  findNodeHandle: jest.fn(),
  Platform: { select: jest.fn(platforms => platforms.ios) },
  NativeModules: { PDFView: { reload: jest.fn() } },
}));

describe('PDFView', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders with default url source type', () => {
    const tree = renderer.create(<PDFVIew resource="http://site.com/pdf" />);

    expect(tree).toMatchSnapshot();
  });

  it('renders base64', () => {
    const tree = renderer.create(<PDFVIew resource="base64" resourceType="base64" />);

    expect(tree).toMatchSnapshot();
  });

  it('invokes onLoad callback', () => {
    const onLoad = jest.fn();
    const tree = renderer.create(<PDFVIew
      onLoad={onLoad}
      resource="base64"
      resourceType="base64"
    />);

    (tree.toJSON(): any).props.onLoad();

    expect(onLoad).toHaveBeenCalledTimes(1);
  });

  it('invokes onError callback', () => {
    const onError = jest.fn();
    const tree = renderer.create(<PDFVIew
      onError={onError}
      resource="base64"
      resourceType="base64"
    />);

    (tree.toJSON(): any).props.onError();

    expect(onError).toHaveBeenCalledTimes(1);
    expect(onError.mock.calls).toMatchSnapshot();
  });

  it('invokes onError callback and provides native error', () => {
    const onError = jest.fn();
    const tree = renderer.create(<PDFVIew
      onError={onError}
      resource="base64"
      resourceType="base64"
    />);

    (tree.toJSON(): any).props.onError({ nativeEvent: { message: 'native error' } });

    expect(onError).toHaveBeenCalledTimes(1);
    expect(onError.mock.calls).toMatchSnapshot();
  });

  it('invokes onPageChanged callback', () => {
    const onPageChanged = jest.fn();
    const tree = renderer.create(<PDFVIew
      onPageChanged={onPageChanged}
      resource="base64"
      resourceType="base64"
    />);

    (tree.toJSON(): any).props.onPageChanged({ nativeEvent: { page: 2, pageCount: 10 } });

    expect(onPageChanged).toHaveBeenCalledTimes(1);
    expect(onPageChanged.mock.calls).toMatchSnapshot();
  });

  it('uses default callbacks', () => {
    const tree = renderer.create(<PDFVIew
      resource="base64"
      resourceType="base64"
    />);

    (tree.toJSON(): any).props.onError();
    (tree.toJSON(): any).props.onLoad();
    (tree.toJSON(): any).props.onPageChanged();
  });

  it('trigger reload', (done) => {
    let pdfRef: any;

    (findNodeHandle: any).mockImplementationOnce(() => 'handle for ref');
    const tree = renderer.create(<PDFVIew
      resource="base64"
      resourceType="base64"
      ref={(ref) => {
        pdfRef = ref;
      }}
    />);

    const component = tree.getInstance();
    const ref = { _name: 'ref_to_the_viewer' };

    // $FlowFixMe: ignore null
    component._setViewRef(ref);

    setTimeout(() => {
      expect(pdfRef.reload).toBeTruthy();
      pdfRef.reload();

      expect(findNodeHandle).toHaveBeenCalledTimes(1);
      expect(NativeModules.PDFView.reload).toHaveBeenCalledTimes(1);
      done();
    });
  });

  it('trigger reload throws exception if findNodeHandle fails to find handle', (done) => {
    let pdfRef: any;

    const tree = renderer.create(<PDFVIew
      resource="base64"
      resourceType="base64"
      ref={(ref) => {
        pdfRef = ref;
      }}
    />);

    const component = tree.getInstance();
    const ref = { _name: 'ref_to_the_viewer' };

    // $FlowFixMe: ignore null
    component._setViewRef(ref);

    setTimeout(async () => {
      expect(pdfRef.reload).toBeTruthy();
      let error;

      try {
        await pdfRef.reload();
      } catch (err) {
        error = err;
      }

      expect(error).toMatchSnapshot();
      done();
    });
  });
});
