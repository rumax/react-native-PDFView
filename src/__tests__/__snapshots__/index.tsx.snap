// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDFView invokes onError callback 1`] = `
Array [
  Array [
    [Error: unknown error],
  ],
]
`;

exports[`PDFView invokes onError callback and provides native error 1`] = `
Array [
  Array [
    Object {
      "message": "native error",
    },
  ],
]
`;

exports[`PDFView invokes onPageChanged callback 1`] = `
Array [
  Array [
    2,
    10,
  ],
]
`;

exports[`PDFView renders base64 1`] = `
<NativeComponent
  enableAnnotations={false}
  fadeInDuration={0}
  fileFrom="bundle"
  onError={[Function]}
  onPageChanged={[Function]}
  onScrolled={[Function]}
  resource="base64"
  resourceType="base64"
  textEncoding="utf-8"
  urlProps={Object {}}
/>
`;

exports[`PDFView renders with default url source type 1`] = `
<NativeComponent
  enableAnnotations={false}
  fadeInDuration={0}
  fileFrom="bundle"
  onError={[Function]}
  onPageChanged={[Function]}
  onScrolled={[Function]}
  resource="http://site.com/pdf"
  resourceType="url"
  textEncoding="utf-8"
  urlProps={Object {}}
/>
`;

exports[`PDFView trigger reload throws exception if findNodeHandle fails to find handle 1`] = `[Error: Cannot find node handles]`;
